lib/extras/packed_image_convert.cc

```diff
@@ -156,5 +163,90 @@ Status ConvertPackedPixelFileToCodecInOut(const PackedPixelFile& ppf,
+  ppf->info.ysize = io.metadata.size.ysize();
+  ppf->info.bits_per_sample = io.metadata.m.bit_depth.bits_per_sample;
+  ppf->info.exponent_bits_per_sample =
+      io.metadata.m.bit_depth.exponent_bits_per_sample;
+
+  ppf->info.alpha_bits = io.metadata.m.GetAlphaBits();
+  ppf->info.alpha_premultiplied = alpha_premultiplied;
+
+  ppf->info.uses_original_profile = !io.metadata.m.xyb_encoded;
+  JXL_ASSERT(0 < io.metadata.m.orientation && io.metadata.m.orientation <= 8);
+  ppf->info.orientation =
+      static_cast<JxlOrientation>(io.metadata.m.orientation);
+
+  // Convert animation metadata
+  JXL_ASSERT(io.frames.size() == 1 || io.metadata.m.have_animation);
+  ppf->info.have_animation = io.metadata.m.have_animation;
+  ppf->info.animation.tps_numerator = io.metadata.m.animation.tps_numerator;
+  ppf->info.animation.tps_denominator = io.metadata.m.animation.tps_denominator;
+  ppf->info.animation.num_loops = io.metadata.m.animation.num_loops;
+
+  // Convert the color encoding
+  // TODO(firsching): do specific ICC profiles need special treatment here?
+  ConvertInternalToExternalColorEncoding(io.metadata.m.color_encoding,
+                                         &ppf->color_encoding);
+  io.metadata.m.color_encoding.IsSRGB();
+
+  // Convert the extra blobs
+  ppf->metadata.exif.assign(io.blobs.exif.begin(), io.blobs.exif.end());
+  ppf->metadata.iptc.assign(io.blobs.iptc.begin(), io.blobs.iptc.end());
+  ppf->metadata.jumbf.assign(io.blobs.jumbf.begin(), io.blobs.jumbf.end());
+  ppf->metadata.xmp.assign(io.blobs.xmp.begin(), io.blobs.xmp.end());
+
+  // Convert the pixels
+  ppf->frames.clear();
+  for (const auto& frame : io.frames) {
+    size_t frame_bits_per_sample = frame.metadata()->bit_depth.bits_per_sample;
+    JXL_ASSERT(frame_bits_per_sample != 0);
+    // It is ok for the frame.color().kNumPlanes to not match the
+    // number of channels on the image.
+    const bool float_out = frame.metadata()->bit_depth.floating_point_sample;
+    const uint32_t num_channels = frame.color().kNumPlanes;
+    JxlPixelFormat format{/*num_channels=*/num_channels,
+                          /*data_type=*/pixel_format.data_type,
+                          /*endianness=*/pixel_format.endianness,
+                          /*align=*/pixel_format.align};
+
+    PackedFrame packed_frame(frame.xsize(), frame.ysize(), format);
+    packed_frame.name = frame.name;
+    packed_frame.frame_info.name_length = frame.name.size();
+    JXL_RETURN_IF_ERROR(ConvertToExternal(
+        frame.Copy(), frame_bits_per_sample, float_out, format.num_channels,
+        format.endianness,
+        /* stride_out=*/packed_frame.color.stride, pool,
+        packed_frame.color.pixels(), packed_frame.color.pixels_size,
+        /*out_callback=*/nullptr, /*out_opaque=*/nullptr,
+        frame.metadata()->GetOrientation()));
+
+    // TODO(firsching): Convert the extra channels. FIXME!
+    JXL_CHECK(frame.extra_channels().empty());
```

Comment:
I think at least alpha should be made to work. Other extra channels can be ignored for now.

Reply:
If I understand correctly, currently the (png, jpg, etc) decoder, that decode images with also as having 4 color channels and 0 extra channels, hence they are using those two fields in `JxlBasicInfo` with a different meaning. 
I agree that alpha should work, and for now, it should work when it comes as interleaved. 
I plan to get some more clarity of what is going on exactly here, and take action in a follow up pull request.

Reply:
`frame.color().kNumPlanes` is always going to be 3, if I'm not mistaken. Alpha is never included in `color()`, and I think it's always going to be RGB, even in the case of grayscale (not sure though).

Reply:
Things are confusing here.

`frame` here is a `CodecInOut` frame, which at least at the moment has the convention that `color()` is always Image3F (so planar float32 RGB) and anything else (including alpha) is in the extra channels.

A `PackedPixelFile` uses a representation similar to what we use in the API (interleaved, arbitrary pixel types, 1 to 4 channels (Y, YA, RGB, RGBA), any other extra channels are separate).

But in this direction, from CodecInOut to PackedPixelFile, the input alpha will always be separate, and it's probably best to return a PackedPixelFile with the alpha interleaved with the RGB because that's what the PNG encoder (and PAM if we add that at some point) will need.

Reply:
There is also `frame.alpha`, right?



> But in this direction, from CodecInOut to PackedPixelFile, the input alpha will always be separate, and it's probably best to return a PackedPixelFile with the alpha interleaved with the RGB because that's what the PNG encoder (and PAM if we add that at some point) will need.
I agree and this is exactly what I'm trying to do. I also think that this works already in the code, because `ConvertToExtrenal` should do the interleaving if alpha is present in the image bundle and the num_channels is provided correctly. 

I'll change the check 
```frame.extra_channels().empty()```, which should have been 
```frame.HasExtraChannels()``` anyhow to check that it has any other extra channels, besides a potential alpha channel. Also I make sure that num_channels is correct.